### PR TITLE
Add debugging mode to conformance test suite executable.

### PR DESCRIPTION
**Pull request by thealmarty**
Added a debugging mode to investigate some failed tests.

I removed the "all or missing" option for adding tests - it doesn't fit with the debugging mode and it adds minimum value because we all use git so we know which files are changed anyway.  

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-conformance/executables/test-utils/Spec.hs [hunk 1]
```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (zliu41) [inline plutus-conformance/executables/test-utils/Spec.hs:1]: This module is using two space indentation. Now that you've renamed this module, I'd just fourmolu it.

File: plutus-conformance/README.md [hunk 2]
```diff
@@ -23,21 +23,21 @@ To update or add test outputs, use the accept test option of the tests. E.g., to
 
 `cabal test haskell-conformance --test-options=--accept`
 
-There is also an executable (`add-test-output`) for adding test output to a specific directory:
+There is also an executable (`test-utils`) for adding/updating test output to a specific directory:
 
 E.g., run
 
-`cabal run add-test-output .uplc plutus-conformance/uplc/ eval -- --missing`
+`cabal run test-utils .uplc plutus-conformance/uplc/ eval -- --missing`
```
Comment (zliu41) [inline plutus-conformance/README.md:30]: The `--missing` option is now removed, right?

File: plutus-conformance/README.md [hunk 3]
```diff
@@ -23,21 +23,21 @@ To update or add test outputs, use the accept test option of the tests. E.g., to
...
 
 to have the executable search for files with extension `.uplc` in the /uplc directory that are missing output files. It will evaluate and create output files for them.
 
 Or, to update and add outputs to all files in the /uplc directory, run
 
-cabal run add-test-output .uplc plutus-conformance/uplc/ eval -- --all
+cabal run test-utils .uplc plutus-conformance/uplc/ eval -- --all
```
Comment (zliu41) [inline plutus-conformance/README.md:36]: ditto
Also this should be put in between backticks.

File: plutus-conformance/executables/test-utils/Spec.hs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  ["This program adds test outputs to specified inputs."
+  , "To run the program, input the following 3 arguments:"
+  , "(1) file extension to be searched"
+  , "(2) directory to be searched"
+  , "(3) the action to run the input files through;"
+  , "eval (for evaluation tests),"
+  , "or typecheck (for typechecking tests),"
+  , "or debug (for debugging failed tests)."
+  , "E.g. run \n"
+  , "cabal run test-utils .uplc plutus-conformance/test-cases/uplc/ eval -- --missing \n"
```
Comment (zliu41) [inline plutus-conformance/executables/test-utils/Spec.hs:89]: ditto

File: plutus-conformance/executables/test-utils/Spec.hs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            GenTestOutput Typecheck ->
+              putStrLn $
+                "typechecking has not been implemented yet." <>
+                  "Only evaluation tests (eval) or debugging (debug) are supported."
+            Debug ->
+              case agdaEvalUplcProgDebug (() <$ pro) of
+                (Right _prog) -> pure ()
+                (Left err)      ->
+                  -- warn the user that the file failed to evaluate
+                  putStrLn $ inputFile <> " failed to evaluate. " <> err
```
Comment (zliu41) [inline plutus-conformance/executables/test-utils/Spec.hs:132]: Aren't some test cases expected to fail? We shouldn't print the debug information in that case.

**Review by zliu41 — CHANGES_REQUESTED** [review]
Is there a JIRA ticket for this PR? I don't think adding a Debug mode to this executable and hardcoding it to use our Agda evaluator is the right approach. Instead I would do

```haskell
- type UplcEvaluator = UplcProg -> Maybe UplcProg
+ type UplcEvaluator = UplcProg -> Either Text UplcProg
```

and print the Text whenever evaluation fails unexpectedly.

Also in the future, it would be nice to open two separate PRs, one for the renaming and one for the rest of the changes.

**Comment by thealmarty** [comment]
> Is there a JIRA ticket for this PR? 

It's kinda a step for PLT-579. I've asked in the stand up, since it didn't take very long I didn't create a ticket for it.

>I don't think adding a Debug mode to this executable and hardcoding it to use our Agda evaluator is the right approach. >Instead I would do
> 
> ```haskell
> - type UplcEvaluator = UplcProg -> Maybe UplcProg
> + type UplcEvaluator = UplcProg -> Either Text UplcProg
> ```
> 
> and print the Text whenever evaluation fails unexpectedly.

This is to debug the expected failure tests so that I can fix them. It can be for unexpectedly failed tests too of course. Also, we've discussed the return type in [this thread](https://github.com/input-output-hk/plutus/pull/4688#discussion_r894770244), we decided `Maybe` is the simplest for users to test their implementation.

> 
> Also in the future, it would be nice to open two separate PRs, one for the renaming and one for the rest of the changes.

OK will do. Sorry for the confusion.

File: plutus-conformance/executables/test-utils/Spec.hs (hunk 5 above)
Comment (thealmarty) [reply plutus-conformance/executables/test-utils/Spec.hs:132]: As mentioned above, the debugger is for those test cases too.

**Comment by thealmarty** [comment]
> > This is to debug the expected failure tests so that I can fix them.
> 
> OK, this wasn't clear to me. 

Sorry about that. I should've mentioned it in the PR description. I mentioned it in the stand up but not everyone gets a record of the verbal one.

>Then my suggestion would be: add an option to print debug info for a specific test case (or test cases). Here's why:
> 
>     * your debug mode, if I understand it correctly, just prints the debug info for _all_ failed evaluations. Failed evaluations fall into two categories: (1) the evaluation is designed to fail, and we know why exactly it fails; (2) we don't know why it fails. There's no need to print debug info for (1). I don't know the ratio of (1) and (2), but I wouldn't be surprised if (1) is the majority.
> 
>     * also, if you just want to fix one single test case, you wouldn't want to re-generate everything every single time.

So you can actually specify the directory to be *the* test case in the script. So you would input the path to be the directory of the test case. Or if you just want to look at e.g. the builtin dir you can do that too. I can make it more clear in the doc/readme.

File: plutus-conformance/src/PlutusConformance/Common.hs [hunk 6]
```diff
@@ -58,7 +61,7 @@ type UplcEvaluator = UplcProg -> Maybe UplcProg
 -- and test cases for directories without.
 discoverTests :: UplcEvaluator -- ^ The evaluator to be tested.
     -> (FilePath -> Bool)
-    -- ^ A function that takes a test name and returns
+    -- ^ A function that takes a test name and returned
```
Comment (zliu41) [inline plutus-conformance/src/PlutusConformance/Common.hs:64]: I guess this is unintentional?

File: plutus-conformance/README.md [hunk 7]
```diff
@@ -23,21 +23,31 @@ To update or add test outputs, use the accept test option of the tests. E.g., to
...
+There is also an executable (`test-utils`) for adding/updating test output to a specific directory:
 
 E.g., run
 
-`cabal run add-test-output .uplc plutus-conformance/uplc/ eval -- --missing`
+`cabal run test-utils .uplc plutus-conformance/test-cases/uplc/evaluation eval`
 
 to have the executable search for files with extension `.uplc` in the /uplc directory that are missing output files. It will evaluate and create output files for them.
 
 Or, to update and add outputs to all files in the /uplc directory, run
```
Comment (zliu41) [inline plutus-conformance/README.md:34]: This should be removed.

**Comment by zliu41** [comment]
> So you can actually specify the directory to be the test case in the script.

Ah, I see.

**Review by zliu41 — COMMENTED** [review]
Looks fine to me but I'd suggest getting a second review from @bezirg or @michaelpj.
I still think printing the debug info from unexpectedly failed test cases is a better approach.

**Review by bezirg — APPROVED** [review]

**Review by zliu41 — APPROVED** [review]
There's still a few "returned", otherwise LGTM

**Comment by thealmarty** [comment]
> There's still a few "returned", otherwise LGTM

OMG thanks so much!
